In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [ ]:
car_df = pd.read_csv('Car Price Prediction.csv')
car_df.head()

car_ID  symboling                   CarName fueltype aspiration doornumber  \
0       1          3        alfa-romero giulia      gas        std        two   
1       2          3       alfa-romero stelvio      gas        std        two   
2       3          1  alfa-romero Quadrifoglio      gas        std        two   
3       4          2               audi 100 ls      gas        std       four   
4       5          2                audi 100ls      gas        std       four   

       carbody drivewheel enginelocation  wheelbase  ...  enginesize  \
0  convertible        rwd          front       88.6  ...         130   
1  convertible        rwd          front       88.6  ...         130   
2    hatchback        rwd          front       94.5  ...         152   
3        sedan        fwd          front       99.8  ...         109   
4        sedan        4wd          front       99.4  ...         136   

   fuelsystem  boreratio  stroke compressionratio horsepower  peakrpm citympg  \
0        mpfi       3.47    2.68              9.0        111     5000      21   
1        mpfi       3.47    2.68              9.0        111     5000      21   
2        mpfi       2.68    3.47              9.0        154     5000      19   
3        mpfi       3.19    3.40             10.0        102     5500      24   
4        mpfi       3.19    3.40              8.0        115     5500      18   

   highwaympg    price  
0          27  13495.0  
1          27  16500.0  
2          26  16500.0  
3          30  13950.0  
4          22  17450.0  

[5 rows x 26 columns]

In [ ]:
x = car_df[['car_ID', 'symboling', 'CarName', 'fueltype', 'aspiration', 'doornumber', 'carbody', 'drivewheel', 'enginelocation', 'wheelbase', 'carlength', 'carwidth', 'carheight', 
            'curbweight', 'enginetype', 'cylindernumber', 'enginesize', 'fuelsystem', 'boreratio', 'stroke', 'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg']]
y = car_df[['price']]

In [ ]:
x.head()

car_ID  symboling                   CarName fueltype aspiration doornumber  \
0       1          3        alfa-romero giulia      gas        std        two   
1       2          3       alfa-romero stelvio      gas        std        two   
2       3          1  alfa-romero Quadrifoglio      gas        std        two   
3       4          2               audi 100 ls      gas        std       four   
4       5          2                audi 100ls      gas        std       four   

       carbody drivewheel enginelocation  wheelbase  ...  cylindernumber  \
0  convertible        rwd          front       88.6  ...            four   
1  convertible        rwd          front       88.6  ...            four   
2    hatchback        rwd          front       94.5  ...             six   
3        sedan        fwd          front       99.8  ...            four   
4        sedan        4wd          front       99.4  ...            five   

   enginesize  fuelsystem  boreratio stroke compressionratio  horsepower  \
0         130        mpfi       3.47   2.68              9.0         111   
1         130        mpfi       3.47   2.68              9.0         111   
2         152        mpfi       2.68   3.47              9.0         154   
3         109        mpfi       3.19   3.40             10.0         102   
4         136        mpfi       3.19   3.40              8.0         115   

  peakrpm  citympg  highwaympg  
0    5000       21          27  
1    5000       21          27  
2    5000       19          26  
3    5500       24          30  
4    5500       18          22  

[5 rows x 25 columns]

In [ ]:
y.head()

price
0  13495.0
1  16500.0
2  16500.0
3  13950.0
4  17450.0

In [ ]:
print(x)

     car_ID  symboling                   CarName fueltype aspiration  \
0         1          3        alfa-romero giulia      gas        std   
1         2          3       alfa-romero stelvio      gas        std   
2         3          1  alfa-romero Quadrifoglio      gas        std   
3         4          2               audi 100 ls      gas        std   
4         5          2                audi 100ls      gas        std   
..      ...        ...                       ...      ...        ...   
200     201         -1           volvo 145e (sw)      gas        std   
201     202         -1               volvo 144ea      gas      turbo   
202     203         -1               volvo 244dl      gas        std   
203     204         -1                 volvo 246   diesel      turbo   
204     205         -1               volvo 264gl      gas      turbo   

    doornumber      carbody drivewheel enginelocation  wheelbase  ...  \
0          two  convertible        rwd          front       88

In [ ]:
print(y)

       price
0    13495.0
1    16500.0
2    16500.0
3    13950.0
4    17450.0
..       ...
200  16845.0
201  19045.0
202  21485.0
203  22470.0
204  22625.0

[205 rows x 1 columns]


In [ ]:
x = pd.get_dummies(x)

In [ ]:
x.head()

car_ID  symboling  wheelbase  carlength  carwidth  carheight  curbweight  \
0       1          3       88.6      168.8      64.1       48.8        2548   
1       2          3       88.6      168.8      64.1       48.8        2548   
2       3          1       94.5      171.2      65.5       52.4        2823   
3       4          2       99.8      176.6      66.2       54.3        2337   
4       5          2       99.4      176.6      66.4       54.3        2824   

   enginesize  boreratio  stroke  ...  cylindernumber_twelve  \
0         130       3.47    2.68  ...                      0   
1         130       3.47    2.68  ...                      0   
2         152       2.68    3.47  ...                      0   
3         109       3.19    3.40  ...                      0   
4         136       3.19    3.40  ...                      0   

   cylindernumber_two  fuelsystem_1bbl  fuelsystem_2bbl  fuelsystem_4bbl  \
0                   0                0                0                0   
1                   0                0                0                0   
2                   0                0                0                0   
3                   0                0                0                0   
4                   0                0                0                0   

   fuelsystem_idi  fuelsystem_mfi  fuelsystem_mpfi  fuelsystem_spdi  \
0               0               0                1                0   
1               0               0                1                0   
2               0               0                1                0   
3               0               0                1                0   
4               0               0                1                0   

   fuelsystem_spfi  
0                0  
1                0  
2                0  
3                0  
4                0  

[5 rows x 200 columns]

In [ ]:
# splitting into training and testing set
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, random_state = 0)

In [ ]:
x_train.head()

car_ID  symboling  wheelbase  carlength  carwidth  carheight  curbweight  \
40       41          0       96.5      175.4      62.5       54.1        2372   
60       61          0       98.8      177.8      66.5       55.5        2410   
56       57          3       95.3      169.0      65.7       49.6        2380   
101     102          0      100.4      181.7      66.5       55.1        3095   
86       87          1       96.3      172.4      65.4       51.6        2405   

     enginesize  boreratio  stroke  ...  cylindernumber_twelve  \
40          110       3.15   3.580  ...                      0   
60          122       3.39   3.390  ...                      0   
56           70       3.33   3.255  ...                      0   
101         181       3.43   3.270  ...                      0   
86          122       3.35   3.460  ...                      0   

     cylindernumber_two  fuelsystem_1bbl  fuelsystem_2bbl  fuelsystem_4bbl  \
40                    0                1                0                0   
60                    0                0                1                0   
56                    1                0                0                1   
101                   0                0                0                0   
86                    0                0                1                0   

     fuelsystem_idi  fuelsystem_mfi  fuelsystem_mpfi  fuelsystem_spdi  \
40                0               0                0                0   
60                0               0                0                0   
56                0               0                0                0   
101               0               0                1                0   
86                0               0                0                0   

     fuelsystem_spfi  
40                 0  
60                 0  
56                 0  
101                0  
86                 0  

[5 rows x 200 columns]

In [ ]:
print(x_test)

     car_ID  symboling  wheelbase  carlength  carwidth  carheight  curbweight  \
52       53          1       93.1      159.1      64.2       54.1        1905   
181     182         -1      104.5      187.8      66.5       54.1        3151   
5         6          2       99.8      177.3      66.3       53.1        2507   
18       19          2       88.4      141.1      60.3       53.2        1488   
188     189          2       97.3      171.7      65.5       55.7        2300   
..      ...        ...        ...        ...       ...        ...         ...   
198     199         -2      104.3      188.8      67.2       56.2        3045   
143     144          0       97.2      172.0      65.4       52.5        2340   
24       25          1       93.7      157.3      63.8       50.6        1967   
30       31          2       86.6      144.6      63.9       50.8        1713   
135     136          2       99.1      186.6      66.5       56.1        2758   

     enginesize  boreratio 

In [ ]:
print(x_train)

     car_ID  symboling  wheelbase  carlength  carwidth  carheight  curbweight  \
40       41          0       96.5      175.4      62.5       54.1        2372   
60       61          0       98.8      177.8      66.5       55.5        2410   
56       57          3       95.3      169.0      65.7       49.6        2380   
101     102          0      100.4      181.7      66.5       55.1        3095   
86       87          1       96.3      172.4      65.4       51.6        2405   
..      ...        ...        ...        ...       ...        ...         ...   
67       68         -1      110.0      190.9      70.3       56.5        3515   
192     193          0      100.4      180.2      66.9       55.1        2579   
117     118          0      108.0      186.7      68.3       56.0        3130   
47       48          0      113.0      199.6      69.6       52.8        4066   
172     173          2       98.4      176.2      65.6       53.0        2975   

     enginesize  boreratio 

In [ ]:
#fitting multiple linear regression to training set
from sklearn.linear_model import LinearRegression
lm = LinearRegression()
lm.fit(x_train, y_train)

LinearRegression()

In [ ]:
print(y_test)

       price
52    6795.0
181  15750.0
5    15250.0
18    5151.0
188   9995.0
..       ...
198  18420.0
143   9960.0
24    6229.0
30    6479.0
135  15510.0

[62 rows x 1 columns]


In [ ]:
print(y_test)

       price
52    6795.0
181  15750.0
5    15250.0
18    5151.0
188   9995.0
..       ...
198  18420.0
143   9960.0
24    6229.0
30    6479.0
135  15510.0

[62 rows x 1 columns]


In [ ]:
# predicting the test set result
pred = lm.predict(x_test)
np.set_printoptions(precision=2)
print(pred)

[[-14183.91]
 [ 63661.73]
 [-13855.5 ]
 [-73516.88]
 [ 56045.83]
 [ 39594.88]
 [ -5429.99]
 [ 34579.91]
 [ 31089.21]
 [-47923.26]
 [ -2362.17]
 [ 63223.94]
 [ 21758.93]
 [  2392.18]
 [-15814.7 ]
 [ -3860.09]
 [  -174.4 ]
 [-14001.56]
 [ 11812.08]
 [  5414.26]
 [  8772.41]
 [ 69206.39]
 [  2025.16]
 [ 23630.63]
 [ 47769.16]
 [ 11253.86]
 [  6947.88]
 [ 66559.98]
 [ 10324.01]
 [ -3382.98]
 [ 45713.6 ]
 [ 45344.24]
 [ 54881.07]
 [  9662.98]
 [  5825.99]
 [ 11048.07]
 [ 39199.68]
 [ 54309.25]
 [-31546.58]
 [ 56710.41]
 [ -5852.97]
 [ 13262.74]
 [ 32421.89]
 [ 97235.43]
 [ 12704.45]
 [ 16161.4 ]
 [-30741.84]
 [ 15764.62]
 [  9608.15]
 [ 46908.9 ]
 [-10836.79]
 [ -1061.35]
 [-23603.27]
 [-11489.35]
 [ 23889.61]
 [ 48829.16]
 [ 24893.12]
 [  9764.17]
 [ 16435.51]
 [-30966.13]
 [ -5781.03]
 [ 46116.13]]


# Building the optimal model using Backward elimination

In [ ]:
import statsmodels.regression.linear_model as sm 
# add a column of ones as integer data type
x = np.append(arr = np.ones((205, 1)).astype(int), values= x, axis = 1)
x_opt = x[:, [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25]]
print(x_opt)

[[  1.   1.   3. ...   0.   0.   0.]
 [  1.   2.   3. ...   0.   0.   0.]
 [  1.   3.   1. ...   0.   0.   0.]
 ...
 [  1. 203.  -1. ...   0.   0.   0.]
 [  1. 204.  -1. ...   0.   0.   0.]
 [  1. 205.  -1. ...   0.   0.   0.]]


In [ ]:
# remove x9
x_opt = x[:, [0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25]]
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.868
Model:                            OLS   Adj. R-squared:                  0.850
Method:                 Least Squares   F-statistic:                     49.25
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           2.71e-66
Time:                        04:46:37   Log-Likelihood:                -1925.0
No. Observations:                 205   AIC:                             3900.
Df Residuals:                     180   BIC:                             3983.
Df Model:                          24                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -5.042e+04   1.68e+04     -3.002      0.003   -8.36e+04   -1.73e+04
x1           -18.0958      4.466     -4.052      0.000     -26.907      -9.284
x2           248.3039    240.961      1.030      0.304    -227.168     723.776
x3           143.1296    109.117      1.312      0.191     -72.184     358.443
x4           -58.5876     55.969     -1.047      0.297    -169.028      51.853
x5           362.8177    269.238      1.348      0.179    -168.451     894.087
x6           254.0050    139.125      1.826      0.070     -20.522     528.532
x7             1.6123      1.744      0.925      0.356      -1.828       5.053
x8           107.7180     14.175      7.599      0.000      79.747     135.689
x9         -3352.0444    772.190     -4.341      0.000   -4875.754   -1828.335
x10          315.5821     82.128      3.843      0.000     153.525     477.639
x11           42.3895     16.209      2.615      0.010      10.405      74.374
x12            1.9485      0.648      3.008      0.003       0.670       3.227
x13         -290.1947    176.608     -1.643      0.102    -638.684      58.294
x14          218.2776    157.336      1.387      0.167     -92.182     528.737
x15        -1104.8052   3134.756     -0.352      0.725   -7290.402    5080.791
x16        -2411.3153   3203.911     -0.753      0.453   -8733.372    3910.741
x17        -1472.9778   3307.687     -0.445      0.657   -7999.808    5053.852
x18         1550.1179   3306.961      0.469      0.640   -4975.279    8075.515
x19          302.7920   3192.986      0.095      0.925   -5997.706    6603.290
x20         -773.1896   2371.069     -0.326      0.745   -5451.856    3905.476
x21         3857.7005   3395.077      1.136      0.257   -2841.571    1.06e+04
x22         -692.7940   3371.740     -0.205      0.837   -7346.016    5960.428
x23         -690.1321   3235.556     -0.213      0.831   -7074.632    5694.367
x24        -1459.2846   3181.192     -0.459      0.647   -7736.510    4817.941
==============================================================================
Omnibus:                       21.201   Durbin-Watson:                   0.985
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               74.032
Skew:                           0.252   Prob(JB):                     8.40e-17
Kurtosis:                       5.901   Cond. No.                     4.50e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.5e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# remove x19
x_opt = x[:, [0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 21, 22, 23, 24, 25]]
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.868
Model:                            OLS   Adj. R-squared:                  0.851
Method:                 Least Squares   F-statistic:                     51.67
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           3.74e-67
Time:                        04:46:37   Log-Likelihood:                -1925.0
No. Observations:                 205   AIC:                             3898.
Df Residuals:                     181   BIC:                             3978.
Df Model:                          23                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -5.058e+04   1.67e+04     -3.037      0.003   -8.34e+04   -1.77e+04
x1           -18.1623      4.398     -4.129      0.000     -26.841      -9.484
x2           250.6139    239.070      1.048      0.296    -221.108     722.336
x3           143.3945    108.782      1.318      0.189     -71.250     358.039
x4           -58.1580     55.633     -1.045      0.297    -167.930      51.614
x5           363.9490    268.236      1.357      0.177    -165.324     893.222
x6           254.8330    138.471      1.840      0.067     -18.391     528.057
x7             1.5916      1.725      0.923      0.357      -1.813       4.996
x8           107.6688     14.127      7.622      0.000      79.795     135.543
x9         -3350.1664    769.820     -4.352      0.000   -4869.142   -1831.190
x10          315.9009     81.834      3.860      0.000     154.429     477.372
x11           42.4869     16.132      2.634      0.009      10.656      74.318
x12            1.9483      0.646      3.016      0.003       0.674       3.223
x13         -289.9622    176.107     -1.647      0.101    -637.449      57.525
x14          217.8356    156.836      1.389      0.167     -91.626     527.297
x15        -1111.6601   3125.331     -0.356      0.722   -7278.429    5055.109
x16        -2415.2447   3194.861     -0.756      0.451   -8719.207    3888.718
x17        -1475.6216   3298.502     -0.447      0.655   -7984.085    5032.842
x18         1547.5406   3297.784      0.469      0.639   -4959.505    8054.586
x19         -789.0581   2358.673     -0.335      0.738   -5443.090    3864.974
x20         3836.3794   3378.338      1.136      0.258   -2829.612    1.05e+04
x21         -711.8958   3356.491     -0.212      0.832   -7334.781    5910.989
x22         -695.0897   3226.265     -0.215      0.830   -7061.018    5670.838
x23        -1473.9461   3168.722     -0.465      0.642   -7726.332    4778.440
==============================================================================
Omnibus:                       21.142   Durbin-Watson:                   0.985
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               73.652
Skew:                           0.251   Prob(JB):                     1.02e-16
Kurtosis:                       5.893   Cond. No.                     4.47e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.47e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# remove x21
x_opt = x[:, [0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 21, 22, 24, 25]]
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.868
Model:                            OLS   Adj. R-squared:                  0.852
Method:                 Least Squares   F-statistic:                     54.30
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           5.12e-68
Time:                        04:46:37   Log-Likelihood:                -1925.1
No. Observations:                 205   AIC:                             3896.
Df Residuals:                     182   BIC:                             3973.
Df Model:                          22                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       -4.95e+04   1.58e+04     -3.131      0.002   -8.07e+04   -1.83e+04
x1           -18.0157      4.332     -4.159      0.000     -26.563      -9.468
x2           248.9665    238.316      1.045      0.298    -221.250     719.183
x3           146.4492    107.542      1.362      0.175     -65.739     358.638
x4           -59.0672     55.322     -1.068      0.287    -168.221      50.087
x5           346.3225    254.367      1.362      0.175    -155.566     848.211
x6           252.1441    137.527      1.833      0.068     -19.208     523.496
x7             1.6119      1.718      0.938      0.349      -1.778       5.002
x8           107.7845     14.079      7.656      0.000      80.005     135.564
x9         -3352.8588    767.693     -4.367      0.000   -4867.582   -1838.135
x10          317.2893     81.358      3.900      0.000     156.764     477.815
x11           42.7814     16.030      2.669      0.008      11.153      74.410
x12            1.9404      0.643      3.017      0.003       0.671       3.210
x13         -289.8081    175.643     -1.650      0.101    -636.367      56.750
x14          217.6046    156.420      1.391      0.166     -91.025     526.234
x15        -1104.9664   3116.961     -0.355      0.723   -7254.993    5045.060
x16        -2417.6549   3186.448     -0.759      0.449   -8704.784    3869.474
x17        -1474.3844   3289.832     -0.448      0.655   -7965.499    5016.730
x18         1548.6313   3289.116      0.471      0.638   -4941.072    8038.334
x19         -718.5400   2328.987     -0.309      0.758   -5313.827    3876.747
x20         3936.3916   3336.479      1.180      0.240   -2646.762    1.05e+04
x21         -665.7908   3214.839     -0.207      0.836   -7008.938    5677.356
x22        -1444.3903   3157.340     -0.457      0.648   -7674.087    4785.306
==============================================================================
Omnibus:                       21.203   Durbin-Watson:                   0.985
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               74.140
Skew:                           0.251   Prob(JB):                     7.96e-17
Kurtosis:                       5.903   Cond. No.                     4.25e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.25e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# remove x21 again
x_opt = x[:, [0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 21, 22, 25]]
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.868
Model:                            OLS   Adj. R-squared:                  0.853
Method:                 Least Squares   F-statistic:                     57.18
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           6.81e-69
Time:                        04:46:37   Log-Likelihood:                -1925.1
No. Observations:                 205   AIC:                             3894.
Df Residuals:                     183   BIC:                             3967.
Df Model:                          21                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -4.952e+04   1.58e+04     -3.141      0.002   -8.06e+04   -1.84e+04
x1           -17.8892      4.278     -4.182      0.000     -26.329      -9.449
x2           254.8391    236.003      1.080      0.282    -210.798     720.476
x3           147.9971    107.001      1.383      0.168     -63.116     359.110
x4           -58.8482     55.167     -1.067      0.287    -167.693      49.996
x5           342.9558    253.183      1.355      0.177    -156.577     842.488
x6           253.0439    137.098      1.846      0.067     -17.452     523.540
x7             1.5804      1.707      0.926      0.356      -1.787       4.948
x8           108.3370     13.788      7.857      0.000      81.133     135.541
x9         -3358.4502    765.209     -4.389      0.000   -4868.218   -1848.683
x10          317.8663     81.097      3.920      0.000     157.861     477.871
x11           42.4913     15.927      2.668      0.008      11.067      73.915
x12            1.9484      0.640      3.042      0.003       0.685       3.212
x13         -289.4795    175.176     -1.653      0.100    -635.104      56.145
x14          217.3142    156.004      1.393      0.165     -90.483     525.112
x15        -1114.3540   3108.471     -0.358      0.720   -7247.404    5018.696
x16        -2382.4682   3173.583     -0.751      0.454   -8643.985    3879.049
x17        -1456.2686   3280.057     -0.444      0.658   -7927.861    5015.323
x18         1566.6206   3279.360      0.478      0.633   -4903.595    8036.836
x19         -700.3980   2321.245     -0.302      0.763   -5280.241    3879.446
x20         3976.1022   3322.242      1.197      0.233   -2578.722    1.05e+04
x21        -1441.1747   3149.034     -0.458      0.648   -7654.257    4771.907
==============================================================================
Omnibus:                       21.306   Durbin-Watson:                   0.986
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               74.295
Skew:                           0.257   Prob(JB):                     7.36e-17
Kurtosis:                       5.904   Cond. No.                     4.25e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.25e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# remove x19 again
x_opt = x[:, [0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 22, 25]]
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.868
Model:                            OLS   Adj. R-squared:                  0.853
Method:                 Least Squares   F-statistic:                     60.34
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           9.04e-70
Time:                        04:46:37   Log-Likelihood:                -1925.1
No. Observations:                 205   AIC:                             3892.
Df Residuals:                     184   BIC:                             3962.
Df Model:                          20                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -4.853e+04   1.54e+04     -3.155      0.002   -7.89e+04   -1.82e+04
x1           -17.6661      4.203     -4.203      0.000     -25.958      -9.374
x2           249.8708    234.846      1.064      0.289    -213.466     713.207
x3           151.0346    106.263      1.421      0.157     -58.615     360.684
x4           -59.6995     54.958     -1.086      0.279    -168.129      48.730
x5           326.9377    246.942      1.324      0.187    -160.265     814.140
x6           248.6058    135.970      1.828      0.069     -19.654     516.866
x7             1.6131      1.699      0.949      0.344      -1.739       4.966
x8           108.3776     13.753      7.880      0.000      81.243     135.512
x9         -3364.2958    763.072     -4.409      0.000   -4869.792   -1858.799
x10          318.5781     80.862      3.940      0.000     159.042     478.114
x11           42.8516     15.843      2.705      0.007      11.595      74.108
x12            1.9386      0.638      3.039      0.003       0.680       3.197
x13         -290.4672    174.712     -1.663      0.098    -635.164      54.230
x14          219.8501    155.392      1.415      0.159     -86.729     526.430
x15        -1100.5603   3100.448     -0.355      0.723   -7217.560    5016.440
x16        -2374.3151   3165.620     -0.750      0.454   -8619.895    3871.265
x17        -1440.0539   3271.506     -0.440      0.660   -7894.541    5014.434
x18         1582.6122   3270.822      0.484      0.629   -4870.526    8035.750
x19         4092.4595   3291.625      1.243      0.215   -2401.721    1.06e+04
x20        -1389.4904   3136.596     -0.443      0.658   -7577.808    4798.827
==============================================================================
Omnibus:                       21.450   Durbin-Watson:                   0.988
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               75.025
Skew:                           0.260   Prob(JB):                     5.11e-17
Kurtosis:                       5.918   Cond. No.                     4.15e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.15e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# remove x15
x_opt = x[:, [0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 13, 14, 15, 17, 18, 19, 22, 25]]
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.868
Model:                            OLS   Adj. R-squared:                  0.854
Method:                 Least Squares   F-statistic:                     63.81
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           1.19e-70
Time:                        04:46:37   Log-Likelihood:                -1925.2
No. Observations:                 205   AIC:                             3890.
Df Residuals:                     185   BIC:                             3957.
Df Model:                          19                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -4.838e+04   1.53e+04     -3.155      0.002   -7.86e+04   -1.81e+04
x1           -17.6367      4.192     -4.207      0.000     -25.907      -9.366
x2           248.6942    234.267      1.062      0.290    -213.484     710.873
x3           151.3393    106.008      1.428      0.155     -57.800     360.479
x4           -60.5770     54.773     -1.106      0.270    -168.637      47.483
x5           326.9616    246.358      1.327      0.186    -159.071     812.995
x6           245.7917    135.417      1.815      0.071     -21.369     512.953
x7             1.6766      1.686      0.995      0.321      -1.649       5.002
x8           108.1182     13.701      7.891      0.000      81.087     135.149
x9         -3371.3324    761.011     -4.430      0.000   -4872.708   -1869.957
x10          318.0487     80.657      3.943      0.000     158.923     477.175
x11           42.7613     15.803      2.706      0.007      11.584      73.939
x12            1.9401      0.636      3.048      0.003       0.684       3.196
x13         -291.0719    174.291     -1.670      0.097    -634.925      52.781
x14          221.0921    154.985      1.427      0.155     -84.674     526.858
x15        -2373.3817   3158.132     -0.752      0.453   -8603.966    3857.202
x16        -1444.8322   3263.742     -0.443      0.659   -7883.770    4994.106
x17         1577.8046   3263.059      0.484      0.629   -4859.787    8015.396
x18         4099.5495   3283.780      1.248      0.213   -2378.922    1.06e+04
x19        -1370.1197   3128.704     -0.438      0.662   -7542.646    4802.407
==============================================================================
Omnibus:                       21.463   Durbin-Watson:                   0.985
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               75.016
Skew:                           0.261   Prob(JB):                     5.13e-17
Kurtosis:                       5.917   Cond. No.                     4.15e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.15e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# remove x19 again
x_opt = x[:, [0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 13, 14, 15, 17, 18, 19, 22]]
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.867
Model:                            OLS   Adj. R-squared:                  0.855
Method:                 Least Squares   F-statistic:                     67.64
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           1.56e-71
Time:                        04:46:37   Log-Likelihood:                -1925.3
No. Observations:                 205   AIC:                             3889.
Df Residuals:                     186   BIC:                             3952.
Df Model:                          18                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -4.834e+04   1.53e+04     -3.159      0.002   -7.85e+04   -1.81e+04
x1           -17.4500      4.161     -4.193      0.000     -25.659      -9.241
x2           242.7988    233.371      1.040      0.300    -217.596     703.193
x3           149.6601    105.708      1.416      0.159     -58.881     358.201
x4           -61.3352     54.626     -1.123      0.263    -169.102      46.432
x5           327.5602    245.819      1.333      0.184    -157.391     812.511
x6           245.1128    135.114      1.814      0.071     -21.440     511.666
x7             1.7569      1.672      1.051      0.295      -1.542       5.056
x8           107.7446     13.645      7.896      0.000      80.826     134.663
x9         -3372.1164    759.354     -4.441      0.000   -4870.169   -1874.064
x10          315.5520     80.280      3.931      0.000     157.175     473.929
x11           42.9824     15.761      2.727      0.007      11.890      74.075
x12            1.9344      0.635      3.047      0.003       0.682       3.187
x13         -287.0585    173.671     -1.653      0.100    -629.677      55.560
x14          221.2825    154.648      1.431      0.154     -83.806     526.371
x15        -2357.1365   3151.046     -0.748      0.455   -8573.520    3859.247
x16        -1415.2242   3255.944     -0.435      0.664   -7838.551    5008.103
x17         1607.2257   3255.272      0.494      0.622   -4814.775    8029.226
x18         4133.1537   3275.743      1.262      0.209   -2329.232    1.06e+04
==============================================================================
Omnibus:                       21.643   Durbin-Watson:                   0.983
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               75.502
Skew:                           0.268   Prob(JB):                     4.03e-17
Kurtosis:                       5.924   Cond. No.                     4.15e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.15e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# remove x16 again
x_opt = x[:, [0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 13, 14, 15, 17, 19, 22]]
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.867
Model:                            OLS   Adj. R-squared:                  0.855
Method:                 Least Squares   F-statistic:                     71.91
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           1.98e-72
Time:                        04:46:37   Log-Likelihood:                -1925.4
No. Observations:                 205   AIC:                             3887.
Df Residuals:                     187   BIC:                             3947.
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -4.966e+04    1.5e+04     -3.318      0.001   -7.92e+04   -2.01e+04
x1           -17.1565      4.097     -4.187      0.000     -25.239      -9.074
x2           242.8508    232.865      1.043      0.298    -216.528     702.230
x3           156.4677    104.314      1.500      0.135     -49.316     362.252
x4           -64.3028     54.080     -1.189      0.236    -170.989      42.383
x5           335.1573    244.664      1.370      0.172    -147.499     817.814
x6           252.2489    133.822      1.885      0.061     -11.745     516.243
x7             1.7047      1.664      1.024      0.307      -1.578       4.988
x8           107.8005     13.615      7.918      0.000      80.942     134.659
x9         -3314.5489    746.091     -4.443      0.000   -4786.386   -1842.712
x10          313.7516     79.999      3.922      0.000     155.934     471.569
x11           43.4768     15.686      2.772      0.006      12.533      74.420
x12            1.9532      0.632      3.090      0.002       0.706       3.200
x13         -286.2075    173.283     -1.652      0.100    -628.048      55.633
x14          222.2072    154.297      1.440      0.152     -82.180     526.594
x15        -2301.2115   3141.583     -0.733      0.465   -8498.709    3896.286
x16         1783.7124   3222.838      0.553      0.581   -4574.080    8141.504
x17         4126.8201   3268.599      1.263      0.208   -2321.247    1.06e+04
==============================================================================
Omnibus:                       21.997   Durbin-Watson:                   0.985
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               75.544
Skew:                           0.289   Prob(JB):                     3.94e-17
Kurtosis:                       5.917   Cond. No.                     4.07e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.07e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# eliminate x16 again
x_opt = x[:, [0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 13, 14, 15, 17, 22]]
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.867
Model:                            OLS   Adj. R-squared:                  0.856
Method:                 Least Squares   F-statistic:                     76.67
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           2.59e-73
Time:                        04:46:37   Log-Likelihood:                -1925.6
No. Observations:                 205   AIC:                             3885.
Df Residuals:                     188   BIC:                             3942.
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -4.818e+04   1.47e+04     -3.278      0.001   -7.72e+04   -1.92e+04
x1           -17.4825      4.047     -4.320      0.000     -25.466      -9.499
x2           242.7967    232.434      1.045      0.298    -215.718     701.312
x3           148.8445    103.210      1.442      0.151     -54.754     352.443
x4           -60.9845     53.648     -1.137      0.257    -166.813      44.844
x5           326.6489    243.730      1.340      0.182    -154.148     807.446
x6           244.2417    132.792      1.839      0.067     -17.712     506.195
x7             1.7632      1.658      1.064      0.289      -1.507       5.033
x8           107.7398     13.589      7.928      0.000      80.933     134.547
x9         -3379.0018    735.586     -4.594      0.000   -4830.064   -1927.939
x10          315.7660     79.769      3.959      0.000     158.409     473.123
x11           42.9204     15.624      2.747      0.007      12.099      73.742
x12            1.9322      0.630      3.068      0.002       0.690       3.175
x13         -287.1670    172.954     -1.660      0.099    -628.348      54.014
x14          221.1726    154.001      1.436      0.153     -82.619     524.965
x15        -2363.5864   3133.763     -0.754      0.452   -8545.444    3818.271
x16         4134.2448   3262.536      1.267      0.207   -2301.638    1.06e+04
==============================================================================
Omnibus:                       21.386   Durbin-Watson:                   0.987
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               73.903
Skew:                           0.265   Prob(JB):                     8.96e-17
Kurtosis:                       5.893   Cond. No.                     4.00e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large,  4e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# eliminate x15
x_opt = x[:, [0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 13, 14, 15, 22]]
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.867
Model:                            OLS   Adj. R-squared:                  0.856
Method:                 Least Squares   F-statistic:                     81.93
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           3.72e-74
Time:                        04:46:38   Log-Likelihood:                -1925.9
No. Observations:                 205   AIC:                             3884.
Df Residuals:                     189   BIC:                             3937.
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -4.918e+04   1.46e+04     -3.363      0.001    -7.8e+04   -2.03e+04
x1           -17.0691      4.005     -4.262      0.000     -24.970      -9.168
x2           254.9143    231.614      1.101      0.272    -201.966     711.795
x3           153.4991    102.908      1.492      0.137     -49.496     356.494
x4           -60.2152     53.577     -1.124      0.262    -165.900      45.470
x5           334.0119    243.256      1.373      0.171    -145.834     813.858
x6           240.1130    132.527      1.812      0.072     -21.310     501.536
x7             1.7252      1.655      1.042      0.299      -1.540       4.990
x8           108.4635     13.540      8.011      0.000      81.755     135.172
x9         -3407.1306    733.801     -4.643      0.000   -4854.623   -1959.638
x10          313.8766     79.639      3.941      0.000     156.782     470.971
x11           41.8625     15.544      2.693      0.008      11.201      72.524
x12            1.9832      0.625      3.171      0.002       0.749       3.217
x13         -279.8671    172.486     -1.623      0.106    -620.113      60.379
x14          216.4939    153.700      1.409      0.161     -86.695     519.683
x15         4150.4955   3258.741      1.274      0.204   -2277.682    1.06e+04
==============================================================================
Omnibus:                       21.851   Durbin-Watson:                   0.993
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               73.787
Skew:                           0.294   Prob(JB):                     9.49e-17
Kurtosis:                       5.880   Cond. No.                     3.98e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.98e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# eliminate x7
x_opt = x[:, [0, 1, 2, 3, 4, 5, 6, 8, 10, 11, 12, 13, 14, 15, 22]]
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.866
Model:                            OLS   Adj. R-squared:                  0.856
Method:                 Least Squares   F-statistic:                     87.67
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           6.65e-75
Time:                        04:46:38   Log-Likelihood:                -1926.5
No. Observations:                 205   AIC:                             3883.
Df Residuals:                     190   BIC:                             3933.
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -5.181e+04   1.44e+04     -3.596      0.000   -8.02e+04   -2.34e+04
x1           -17.2609      4.002     -4.313      0.000     -25.155      -9.367
x2           256.0674    231.664      1.105      0.270    -200.896     713.031
x3           169.6990    101.750      1.668      0.097     -31.007     370.405
x4           -45.4650     51.686     -0.880      0.380    -147.417      56.487
x5           374.4557    240.196      1.559      0.121     -99.339     848.250
x6           250.7204    132.166      1.897      0.059      -9.981     511.422
x7           112.7640     12.899      8.742      0.000      87.321     138.207
x8         -3389.1372    733.765     -4.619      0.000   -4836.509   -1941.765
x9           346.7589     73.140      4.741      0.000     202.487     491.030
x10           46.6208     14.861      3.137      0.002      17.306      75.935
x11            1.8546      0.613      3.024      0.003       0.645       3.064
x12         -275.3344    172.471     -1.596      0.112    -615.538      64.869
x13          176.5250    148.874      1.186      0.237    -117.133     470.183
x14         3937.3074   3253.056      1.210      0.228   -2479.437    1.04e+04
==============================================================================
Omnibus:                       21.570   Durbin-Watson:                   1.033
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               70.187
Skew:                           0.304   Prob(JB):                     5.74e-16
Kurtosis:                       5.801   Cond. No.                     3.51e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.51e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# eliminate x4
x_opt = x[:, [0, 1, 2, 3, 5, 6, 8, 10, 11, 12, 13, 14, 15, 22]]
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.865
Model:                            OLS   Adj. R-squared:                  0.856
Method:                 Least Squares   F-statistic:                     94.46
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           9.82e-76
Time:                        04:46:38   Log-Likelihood:                -1926.9
No. Observations:                 205   AIC:                             3882.
Df Residuals:                     191   BIC:                             3928.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -5.168e+04   1.44e+04     -3.589      0.000   -8.01e+04   -2.33e+04
x1           -17.9351      3.926     -4.569      0.000     -25.678     -10.192
x2           255.3624    231.525      1.103      0.271    -201.312     712.037
x3           128.5523     90.307      1.423      0.156     -49.576     306.680
x4           330.4286    234.785      1.407      0.161    -132.675     793.532
x5           222.7426    128.206      1.737      0.084     -30.139     475.624
x6           110.8022     12.697      8.727      0.000      85.758     135.846
x7         -3376.7872    733.196     -4.606      0.000   -4822.989   -1930.586
x8           336.4240     72.148      4.663      0.000     194.115     478.733
x9            47.5821     14.812      3.212      0.002      18.365      76.799
x10            1.8716      0.613      3.055      0.003       0.663       3.080
x11         -226.6065    163.236     -1.388      0.167    -548.583      95.370
x12          156.7765    147.084      1.066      0.288    -133.342     446.895
x13         3752.1724   3244.317      1.157      0.249   -2647.120    1.02e+04
==============================================================================
Omnibus:                       20.104   Durbin-Watson:                   1.011
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               63.064
Skew:                           0.276   Prob(JB):                     2.02e-14
Kurtosis:                       5.661   Cond. No.                     3.51e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.51e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# eliminate x12
x_opt = x[:, [0, 1, 2, 3, 5, 6, 8, 10, 11, 12, 13, 14, 22]]
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.865
Model:                            OLS   Adj. R-squared:                  0.856
Method:                 Least Squares   F-statistic:                     102.2
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           1.66e-76
Time:                        04:46:38   Log-Likelihood:                -1927.5
No. Observations:                 205   AIC:                             3881.
Df Residuals:                     192   BIC:                             3924.
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -5.054e+04   1.44e+04     -3.518      0.001   -7.89e+04   -2.22e+04
x1           -17.8526      3.926     -4.547      0.000     -25.597     -10.109
x2           272.0737    231.076      1.177      0.240    -183.699     727.847
x3           111.9655     88.988      1.258      0.210     -63.554     287.485
x4           344.5901    234.491      1.470      0.143    -117.920     807.100
x5           229.5708    128.091      1.792      0.075     -23.076     482.218
x6           108.2924     12.481      8.677      0.000      83.675     132.910
x7         -3293.7959    729.309     -4.516      0.000   -4732.282   -1855.310
x8           331.2241     72.008      4.600      0.000     189.195     473.253
x9            49.9940     14.644      3.414      0.001      21.111      78.877
x10            1.8435      0.612      3.011      0.003       0.636       3.051
x11          -68.4994     68.166     -1.005      0.316    -202.950      65.952
x12         3305.8789   3218.326      1.027      0.306   -3041.936    9653.694
==============================================================================
Omnibus:                       20.787   Durbin-Watson:                   0.993
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               67.063
Skew:                           0.283   Prob(JB):                     2.74e-15
Kurtosis:                       5.744   Cond. No.                     3.50e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.5e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# eliminate x11 
x_opt = x[:, [0, 1, 2, 3, 5, 6, 8, 10, 11, 12, 13, 22]]
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.864
Model:                            OLS   Adj. R-squared:                  0.856
Method:                 Least Squares   F-statistic:                     111.4
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           2.54e-77
Time:                        04:46:38   Log-Likelihood:                -1928.1
No. Observations:                 205   AIC:                             3880.
Df Residuals:                     193   BIC:                             3920.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -5.756e+04   1.25e+04     -4.586      0.000   -8.23e+04   -3.28e+04
x1           -17.9600      3.925     -4.576      0.000     -25.701     -10.219
x2           332.6764    223.073      1.491      0.138    -107.297     772.650
x3           137.2882     85.348      1.609      0.109     -31.046     305.622
x4           375.6937    232.446      1.616      0.108     -82.766     834.154
x5           239.6644    127.700      1.877      0.062     -12.203     491.531
x6           106.6308     12.371      8.619      0.000      82.230     131.031
x7         -3301.8721    729.283     -4.528      0.000   -4740.260   -1863.484
x8           299.0562     64.503      4.636      0.000     171.835     426.278
x9            57.3538     12.681      4.523      0.000      32.343      82.365
x10            1.8354      0.612      2.998      0.003       0.628       3.043
x11         3164.6040   3215.335      0.984      0.326   -3177.104    9506.312
==============================================================================
Omnibus:                       20.743   Durbin-Watson:                   1.001
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               64.395
Skew:                           0.302   Prob(JB):                     1.04e-14
Kurtosis:                       5.678   Cond. No.                     3.06e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.06e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# eliminate x11 again
x_opt = x[:, [0, 1, 2, 3, 5, 6, 8, 10, 11, 12, 13]]
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.863
Model:                            OLS   Adj. R-squared:                  0.856
Method:                 Least Squares   F-statistic:                     122.4
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           3.60e-78
Time:                        04:46:38   Log-Likelihood:                -1928.6
No. Observations:                 205   AIC:                             3879.
Df Residuals:                     194   BIC:                             3916.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -6.082e+04   1.21e+04     -5.025      0.000   -8.47e+04   -3.69e+04
x1           -18.4804      3.889     -4.752      0.000     -26.150     -10.811
x2           336.2911    223.025      1.508      0.133    -103.573     776.155
x3           129.0960     84.934      1.520      0.130     -38.417     296.609
x4           431.7137    225.351      1.916      0.057     -12.739     876.166
x5           251.7910    127.094      1.981      0.049       1.128     502.454
x6           104.8536     12.238      8.568      0.000      80.717     128.990
x7         -3280.5966    728.904     -4.501      0.000   -4718.189   -1843.004
x8           294.7632     64.350      4.581      0.000     167.847     421.679
x9            57.6922     12.675      4.552      0.000      32.693      82.691
x10            1.8270      0.612      2.985      0.003       0.620       3.034
==============================================================================
Omnibus:                       20.196   Durbin-Watson:                   1.009
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               60.295
Skew:                           0.305   Prob(JB):                     8.07e-14
Kurtosis:                       5.586   Cond. No.                     2.95e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.95e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# eliminate x2
x_opt = x[:, [0, 1, 3, 5, 6, 8, 10, 11, 12, 13]]
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.862
Model:                            OLS   Adj. R-squared:                  0.855
Method:                 Least Squares   F-statistic:                     134.9
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           9.25e-79
Time:                        04:46:38   Log-Likelihood:                -1929.8
No. Observations:                 205   AIC:                             3880.
Df Residuals:                     195   BIC:                             3913.
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -5.827e+04    1.2e+04     -4.846      0.000    -8.2e+04   -3.46e+04
x1           -18.6313      3.900     -4.777      0.000     -26.323     -10.939
x2            81.1464     79.013      1.027      0.306     -74.684     236.977
x3           493.4651    222.321      2.220      0.028      55.002     931.928
x4           216.9948    125.389      1.731      0.085     -30.298     464.288
x5           104.6769     12.277      8.526      0.000      80.464     128.890
x6         -3250.0700    730.998     -4.446      0.000   -4691.747   -1808.393
x7           296.5721     64.549      4.595      0.000     169.268     423.876
x8            58.7840     12.696      4.630      0.000      33.745      83.823
x9             1.8414      0.614      2.999      0.003       0.630       3.052
==============================================================================
Omnibus:                       20.345   Durbin-Watson:                   1.041
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               66.907
Skew:                           0.257   Prob(JB):                     2.96e-15
Kurtosis:                       5.751   Cond. No.                     2.92e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.92e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# eliminate x2 again
x_opt = x[:, [0, 1, 5, 6, 8, 10, 11, 12, 13]]
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.861
Model:                            OLS   Adj. R-squared:                  0.855
Method:                 Least Squares   F-statistic:                     151.6
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           1.23e-79
Time:                        04:46:38   Log-Likelihood:                -1930.3
No. Observations:                 205   AIC:                             3879.
Df Residuals:                     196   BIC:                             3909.
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -6.423e+04   1.05e+04     -6.099      0.000    -8.5e+04   -4.35e+04
x1           -18.5315      3.899     -4.752      0.000     -26.222     -10.841
x2           644.6398    166.633      3.869      0.000     316.017     973.263
x3           290.2493    103.139      2.814      0.005      86.845     493.653
x4           107.2507     12.020      8.922      0.000      83.545     130.957
x5         -3187.8138    728.582     -4.375      0.000   -4624.680   -1750.947
x6           291.9672     64.402      4.534      0.000     164.957     418.977
x7            56.1028     12.426      4.515      0.000      31.597      80.609
x8             1.8147      0.614      2.958      0.003       0.605       3.025
==============================================================================
Omnibus:                       19.243   Durbin-Watson:                   1.029
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               62.889
Skew:                           0.219   Prob(JB):                     2.21e-14
Kurtosis:                       5.678   Cond. No.                     2.56e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.56e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
x_opt_df = pd.DataFrame(x_opt)
print(x_opt_df)

       0      1     2     3      4     5     6      7       8
0    1.0    1.0  64.1  48.8  130.0  2.68   9.0  111.0  5000.0
1    1.0    2.0  64.1  48.8  130.0  2.68   9.0  111.0  5000.0
2    1.0    3.0  65.5  52.4  152.0  3.47   9.0  154.0  5000.0
3    1.0    4.0  66.2  54.3  109.0  3.40  10.0  102.0  5500.0
4    1.0    5.0  66.4  54.3  136.0  3.40   8.0  115.0  5500.0
..   ...    ...   ...   ...    ...   ...   ...    ...     ...
200  1.0  201.0  68.9  55.5  141.0  3.15   9.5  114.0  5400.0
201  1.0  202.0  68.8  55.5  141.0  3.15   8.7  160.0  5300.0
202  1.0  203.0  68.9  55.5  173.0  2.87   8.8  134.0  5500.0
203  1.0  204.0  68.9  55.5  145.0  3.40  23.0  106.0  4800.0
204  1.0  205.0  68.9  55.5  141.0  3.15   9.5  114.0  5400.0

[205 rows x 9 columns]


In [ ]:
x = x_opt[:, :9]
print(x)

[[1.00e+00 1.00e+00 6.41e+01 ... 9.00e+00 1.11e+02 5.00e+03]
 [1.00e+00 2.00e+00 6.41e+01 ... 9.00e+00 1.11e+02 5.00e+03]
 [1.00e+00 3.00e+00 6.55e+01 ... 9.00e+00 1.54e+02 5.00e+03]
 ...
 [1.00e+00 2.03e+02 6.89e+01 ... 8.80e+00 1.34e+02 5.50e+03]
 [1.00e+00 2.04e+02 6.89e+01 ... 2.30e+01 1.06e+02 4.80e+03]
 [1.00e+00 2.05e+02 6.89e+01 ... 9.50e+00 1.14e+02 5.40e+03]]


In [ ]:
y = y.to_numpy()
print(y)

[[13495.  ]
 [16500.  ]
 [16500.  ]
 [13950.  ]
 [17450.  ]
 [15250.  ]
 [17710.  ]
 [18920.  ]
 [23875.  ]
 [17859.17]
 [16430.  ]
 [16925.  ]
 [20970.  ]
 [21105.  ]
 [24565.  ]
 [30760.  ]
 [41315.  ]
 [36880.  ]
 [ 5151.  ]
 [ 6295.  ]
 [ 6575.  ]
 [ 5572.  ]
 [ 6377.  ]
 [ 7957.  ]
 [ 6229.  ]
 [ 6692.  ]
 [ 7609.  ]
 [ 8558.  ]
 [ 8921.  ]
 [12964.  ]
 [ 6479.  ]
 [ 6855.  ]
 [ 5399.  ]
 [ 6529.  ]
 [ 7129.  ]
 [ 7295.  ]
 [ 7295.  ]
 [ 7895.  ]
 [ 9095.  ]
 [ 8845.  ]
 [10295.  ]
 [12945.  ]
 [10345.  ]
 [ 6785.  ]
 [ 8916.5 ]
 [ 8916.5 ]
 [11048.  ]
 [32250.  ]
 [35550.  ]
 [36000.  ]
 [ 5195.  ]
 [ 6095.  ]
 [ 6795.  ]
 [ 6695.  ]
 [ 7395.  ]
 [10945.  ]
 [11845.  ]
 [13645.  ]
 [15645.  ]
 [ 8845.  ]
 [ 8495.  ]
 [10595.  ]
 [10245.  ]
 [10795.  ]
 [11245.  ]
 [18280.  ]
 [18344.  ]
 [25552.  ]
 [28248.  ]
 [28176.  ]
 [31600.  ]
 [34184.  ]
 [35056.  ]
 [40960.  ]
 [45400.  ]
 [16503.  ]
 [ 5389.  ]
 [ 6189.  ]
 [ 6669.  ]
 [ 7689.  ]
 [ 9959.  ]
 [ 8499.  ]
 [12629.  ]
 [14

# Resplit the new Dataset into Training set and Test set

In [ ]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, random_state = 1)

In [ ]:
print(x_train)

[[1.00e+00 1.25e+02 6.63e+01 ... 7.00e+00 1.45e+02 5.00e+03]
 [1.00e+00 1.82e+02 6.65e+01 ... 9.20e+00 1.56e+02 5.20e+03]
 [1.00e+00 1.55e+02 6.36e+01 ... 9.00e+00 6.20e+01 4.80e+03]
 ...
 [1.00e+00 7.30e+01 7.05e+01 ... 8.30e+00 1.55e+02 4.75e+03]
 [1.00e+00 1.41e+02 6.38e+01 ... 8.70e+00 7.30e+01 4.40e+03]
 [1.00e+00 3.80e+01 6.52e+01 ... 9.00e+00 8.60e+01 5.80e+03]]


In [ ]:
print(y_train)

[[12764.  ]
 [15750.  ]
 [ 7898.  ]
 [ 6695.  ]
 [ 7299.  ]
 [ 9960.  ]
 [19699.  ]
 [11850.  ]
 [10345.  ]
 [36880.  ]
 [15250.  ]
 [ 5348.  ]
 [18950.  ]
 [35550.  ]
 [ 9298.  ]
 [18344.  ]
 [17075.  ]
 [ 7395.  ]
 [ 7349.  ]
 [ 7609.  ]
 [ 5499.  ]
 [11549.  ]
 [ 8921.  ]
 [20970.  ]
 [34028.  ]
 [14399.  ]
 [15985.  ]
 [ 7788.  ]
 [ 8949.  ]
 [ 8499.  ]
 [ 9995.  ]
 [12940.  ]
 [21105.  ]
 [12629.  ]
 [15998.  ]
 [ 7975.  ]
 [ 8916.5 ]
 [13200.  ]
 [ 7295.  ]
 [ 9538.  ]
 [13295.  ]
 [ 7898.  ]
 [16900.  ]
 [ 6918.  ]
 [21485.  ]
 [ 5572.  ]
 [18420.  ]
 [16503.  ]
 [ 5572.  ]
 [ 6649.  ]
 [ 8495.  ]
 [16500.  ]
 [31600.  ]
 [ 6989.  ]
 [ 8778.  ]
 [17710.  ]
 [ 8449.  ]
 [13495.  ]
 [ 7738.  ]
 [ 6189.  ]
 [16695.  ]
 [18280.  ]
 [10945.  ]
 [16845.  ]
 [ 9279.  ]
 [11595.  ]
 [11048.  ]
 [45400.  ]
 [ 6849.  ]
 [ 8013.  ]
 [ 9279.  ]
 [22470.  ]
 [16430.  ]
 [ 5399.  ]
 [14869.  ]
 [ 8358.  ]
 [ 9549.  ]
 [22018.  ]
 [ 7957.  ]
 [32528.  ]
 [17859.17]
 [ 9988.  ]
 [17199.  ]
 [ 9

# Retrain multiple linear regression model on training set

In [ ]:
from sklearn.linear_model import LinearRegression
car_regressor = LinearRegression()
car_regressor.fit(x_train, y_train)

LinearRegression()

# Retest the model on the Test set by Repredicting the Test set results

In [ ]:
car_y_pred = car_regressor.predict(x_test) 
print(car_y_pred)

[[ 6463.73]
 [ 6510.96]
 [ 4498.  ]
 [ 7238.61]
 [ 9840.35]
 [25750.52]
 [ 6413.7 ]
 [ 1407.28]
 [ 4477.61]
 [11674.7 ]
 [16889.86]
 [12220.65]
 [16227.38]
 [11144.94]
 [15523.56]
 [11877.12]
 [ 6404.48]
 [ 8772.77]
 [16919.4 ]
 [10684.3 ]
 [ 7786.09]
 [ 8063.6 ]
 [14732.46]
 [ 6591.76]
 [11897.51]
 [ 8752.38]
 [14525.64]
 [ 9131.16]
 [ 4953.87]
 [ 9462.18]
 [ 4848.94]
 [13328.22]
 [28829.75]
 [10763.85]
 [18499.68]
 [ 9171.94]
 [ 8390.97]
 [30235.72]
 [10314.01]
 [16305.31]
 [ 7567.81]
 [21401.  ]
 [ 9837.99]
 [10636.46]
 [ 8374.07]
 [ 6474.88]
 [40448.58]
 [ 7014.97]
 [17976.03]
 [22311.79]
 [25094.83]
 [13772.39]
 [19133.23]
 [ 5178.97]
 [11044.11]
 [ 7748.41]
 [ 7320.21]
 [11432.  ]
 [19390.66]
 [ 4828.55]
 [ 7471.4 ]
 [ 7724.92]]


In [ ]:
print(y_test)

[[ 6669. ]
 [ 7999. ]
 [ 6338. ]
 [ 8916.5]
 [10295. ]
 [25552. ]
 [ 8249. ]
 [ 5151. ]
 [ 6488. ]
 [10245. ]
 [17450. ]
 [17669. ]
 [16515. ]
 [ 9095. ]
 [12964. ]
 [11199. ]
 [ 9980. ]
 [ 6529. ]
 [18150. ]
 [11259. ]
 [ 8948. ]
 [ 6855. ]
 [16925. ]
 [ 7957. ]
 [ 9989. ]
 [ 7129. ]
 [14489. ]
 [ 7995. ]
 [ 6229. ]
 [ 7295. ]
 [ 8058. ]
 [ 8921. ]
 [41315. ]
 [ 9295. ]
 [17950. ]
 [ 7775. ]
 [ 8558. ]
 [32250. ]
 [15645. ]
 [22625. ]
 [ 5118. ]
 [18399. ]
 [ 7099. ]
 [10198. ]
 [ 8195. ]
 [ 7799. ]
 [40960. ]
 [ 6095. ]
 [18150. ]
 [13499. ]
 [28176. ]
 [11900. ]
 [24565. ]
 [ 9258. ]
 [ 8845. ]
 [ 6295. ]
 [11248. ]
 [ 8845. ]
 [13860. ]
 [ 8238. ]
 [11845. ]
 [10898. ]]


# Compare the test set predictions with actual values

In [ ]:
car_y_pred = car_regressor.predict(x_test) 
np.set_printoptions(precision=2)
print(np.concatenate((car_y_pred.reshape(len(car_y_pred), 1), y_test.reshape(len(y_test), 1)), 1))

[[ 6463.73  6669.  ]
 [ 6510.96  7999.  ]
 [ 4498.    6338.  ]
 [ 7238.61  8916.5 ]
 [ 9840.35 10295.  ]
 [25750.52 25552.  ]
 [ 6413.7   8249.  ]
 [ 1407.28  5151.  ]
 [ 4477.61  6488.  ]
 [11674.7  10245.  ]
 [16889.86 17450.  ]
 [12220.65 17669.  ]
 [16227.38 16515.  ]
 [11144.94  9095.  ]
 [15523.56 12964.  ]
 [11877.12 11199.  ]
 [ 6404.48  9980.  ]
 [ 8772.77  6529.  ]
 [16919.4  18150.  ]
 [10684.3  11259.  ]
 [ 7786.09  8948.  ]
 [ 8063.6   6855.  ]
 [14732.46 16925.  ]
 [ 6591.76  7957.  ]
 [11897.51  9989.  ]
 [ 8752.38  7129.  ]
 [14525.64 14489.  ]
 [ 9131.16  7995.  ]
 [ 4953.87  6229.  ]
 [ 9462.18  7295.  ]
 [ 4848.94  8058.  ]
 [13328.22  8921.  ]
 [28829.75 41315.  ]
 [10763.85  9295.  ]
 [18499.68 17950.  ]
 [ 9171.94  7775.  ]
 [ 8390.97  8558.  ]
 [30235.72 32250.  ]
 [10314.01 15645.  ]
 [16305.31 22625.  ]
 [ 7567.81  5118.  ]
 [21401.   18399.  ]
 [ 9837.99  7099.  ]
 [10636.46 10198.  ]
 [ 8374.07  8195.  ]
 [ 6474.88  7799.  ]
 [40448.58 40960.  ]
 [ 7014.97  6

In [ ]:
car_regressor.score(x_train, y_train)

0.8703278228174915

In [ ]:
# predictions were way off given train set was 70% and test set was 30%
car_regressor.score(x_test, y_test)

0.8285110694918039

# using multiple linear regresison to make a single prediction

In [ ]:
car_regressor.predict([[1, 1, 64.1, 48.8, 130, 2.68, 9, 111, 50000]])

array([[117273.29]])

# Display the final regression equation y = b0 + b1 x1 + b2 x2 + ... with the final values of the coefficients?

In [ ]:
print(pd.DataFrame(car_regressor.coef_))

     0          1           2           3           4            5  \
0  0.0 -20.392067  581.946064  384.313706  113.684931 -3091.299902   

            6          7         8  
0  344.876352  51.368919  2.293731  


In [ ]:
print(car_regressor.intercept_)

[-68750.32]


#  the equation of tis multiple linear regression model is: 

price = - 20.4 x car_Id + 581.9 x carwidth + 384. 3 x carheight + 113.7 x enginesize - 3091.3 x stroke + 344.9 x compressionratio + 51.4 x horsepower + 2.29 x peakrpm - 0 